hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/ListActionTaker.java

```diff
@@ -0,0 +1,178 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.fs.azurebfs.services;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.fs.azurebfs.AbfsConfiguration;
+import org.apache.hadoop.fs.azurebfs.contracts.exceptions.AzureBlobFileSystemException;
+import org.apache.hadoop.fs.azurebfs.contracts.services.BlobListResultSchema;
+import org.apache.hadoop.fs.azurebfs.contracts.services.ListResultEntrySchema;
+import org.apache.hadoop.fs.azurebfs.contracts.services.ListResultSchema;
+import org.apache.hadoop.fs.azurebfs.utils.TracingContext;
+
+import static org.apache.hadoop.fs.azurebfs.constants.AbfsHttpConstants.ROOT_PATH;
+
+/**
+ * ListActionTaker is an abstract class that provides a way to list the paths
+ * recursively and take action on each path. The implementations of this class
+ * should provide the action to be taken on each listed path.
+ */
+public abstract class ListActionTaker {
+
+  protected final Path path;
+
+  protected final AbfsBlobClient abfsClient;
+
+  protected final TracingContext tracingContext;
+
+  private final ExecutorService executorService;
+
+  private final int maxConsumptionParallelism;
+
+  private final AtomicBoolean producerThreadToBeStopped = new AtomicBoolean(
+      false);
+
+  public ListActionTaker(Path path,
+      AbfsClient abfsClient,
+      int maxConsumptionParallelism,
+      TracingContext tracingContext) {
+    this.path = path;
+    this.abfsClient = (AbfsBlobClient) abfsClient;
+    this.tracingContext = tracingContext;
+    this.maxConsumptionParallelism = maxConsumptionParallelism;
+    executorService = Executors.newFixedThreadPool(maxConsumptionParallelism);
+  }
+
+  abstract boolean takeAction(Path path) throws IOException;
+
+  private boolean takeAction(List<Path> paths) throws IOException {
+    List<Future<Boolean>> futureList = new ArrayList<>();
+    for (Path path : paths) {
+      Future<Boolean> future = executorService.submit(() -> {
+        return takeAction(path);
+      });
+      futureList.add(future);
+    }
+
+    IOException executionException = null;
+    boolean actionResult = true;
+    for (Future<Boolean> future : futureList) {
+      try {
+        Boolean result = future.get();
+        if (!result) {
+          actionResult = false;
+        }
+      } catch (InterruptedException ignored) {
+
+      } catch (ExecutionException e) {
+        executionException = (IOException) e.getCause();
+      }
+    }
+    if (executionException != null) {
+      throw executionException;
```

Comment:
why should we wait for the loop to complete here and not throw exception as soon as it is caught ?

Reply:
There are parallel threads who are doing the action. It is true that if we throw exception from here, it will shutdown the executorservice and those threads would stop. But the intrusive issue is that the operation needs to be atomic. 

Lets say one thread has done copy, but before delete, the executorService gets stopped and the deletion part is missed, and the blob is both at source and destination. 

One thing to not here is the parallelism is a config number probably small (may be max to 100 (depends on CPU cores etc. ), default is 5). So, if we are waiting, we are waiting for small number of threads to get over.


@anmolanmol1234, let me please know what you feel here, we can change the heuristic if you feel so. But, I am very much inclined to keep operation atomic.

Good point!